xtra-libp2p-offer/src/current/maker.rs

```diff
@@ -0,0 +1,139 @@
+use async_trait::async_trait;
+use std::collections::HashSet;
+use std::time::Duration;
+use tokio_extras::spawn_fallible;
+use tracing::Instrument;
+use xtra_libp2p::endpoint;
+use xtra_libp2p::libp2p::PeerId;
+use xtra_libp2p::Endpoint;
+use xtra_libp2p::GetConnectionStats;
+use xtra_libp2p::OpenSubstream;
+use xtra_productivity::xtra_productivity;
+
+pub struct Actor {
+    endpoint: xtra::Address<Endpoint>,
+    connected_peers: HashSet<PeerId>,
+    latest_offers: Vec<model::Offer>,
+}
+
+impl Actor {
+    pub fn new(endpoint: xtra::Address<Endpoint>) -> Self {
+        Self {
+            endpoint,
+            connected_peers: HashSet::default(),
+            latest_offers: Vec::new(),
+        }
+    }
+
+    async fn send_offers(&self, peer_id: PeerId, ctx: &mut xtra::Context<Self>) {
+        let endpoint = self.endpoint.clone();
+        let offers = self.latest_offers.clone();
+
+        let span = tracing::debug_span!("Send offers", %peer_id).or_current();
+        let task = async move {
+            let stream = endpoint
+                .send(OpenSubstream::single_protocol(peer_id, PROTOCOL))
+                .await??
+                .await?;
+
+            protocol::send(stream, offers.into()).await?;
+
+            anyhow::Ok(())
+        };
+
+        let err_handler = move |e: anyhow::Error| async move {
+            if let Some(xtra_libp2p::Error::NegotiationFailed(
+                xtra_libp2p::NegotiationError::Failed,
+            )) = e.downcast_ref::<xtra_libp2p::Error>()
+            {
+                // It's normal to disagree on the protocols now that we broadcast on both versions
+                // to _all_ our peers
+            } else {
+                tracing::warn!(%peer_id, "Failed to send offers: {e:#}")
+            }
+        };
+
+        let this = ctx.address().expect("self to be alive");
+        spawn_fallible(&this, task.instrument(span), err_handler);
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle(&mut self, msg: NewOffers, ctx: &mut xtra::Context<Self>) {
+        self.latest_offers = msg.0;
+
+        let quiet = quiet_spans::sometimes_quiet_children();
+        for peer_id in self.connected_peers.iter().copied() {
+            self.send_offers(peer_id, ctx)
+                .instrument(quiet.in_scope(|| {
+                    tracing::debug_span!("Broadcast offers to taker (libp2p)").or_current()
+                }))
+                .await
+        }
+    }
+
+    async fn handle(&mut self, _: GetLatestOffers) -> Vec<model::Offer> {
+        self.latest_offers.clone()
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle_connection_established(
+        &mut self,
+        msg: endpoint::ConnectionEstablished,
+        ctx: &mut xtra::Context<Self>,
+    ) {
+        tracing::trace!("Adding newly established connection: {:?}", msg.peer_id);
+        self.connected_peers.insert(msg.peer_id);
+        self.send_offers(msg.peer_id, ctx).await;
+    }
```

Comment:
I've noticed that this is not exactly what we want. Here the maker is only sharing with the taker the very last offers sent over the wire.

I think we want to be able to tell them about all the relevant offers that they have missed. To do so we would need to introduce more complex state in the maker actor i.e. a `HashMap<(ContractSymbol, Position), Offer>`. I spent some time trying to work on this, which also motivated me to model the wire message similarly (as you suggested above, @da-kami).

It gets quite complicated and I don't want to spend more time on this until we have ETHUSD almost working. To move forward I would just create a ticket and add it to the milestone. When we get to tackle it we can decide if we want to do some more remodelling depending on how much time we have.

And I've stashed my changes locally in case we want to revisit them specifically.

Reply:
Fair enough!